File: Stack/연습문제/Calc.java
package basic;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Scanner;
import java.util.Stack;

import javax.naming.event.ObjectChangeListener;

public class Calc {
	private Object c;
	
	/**
	 * 문자열로 들어온 수식을 ArrayList로 분할합니다.
	 * @param expr
	 * @return 수식을 피연산자,연산자로 분할한 ArrayList
	 */
	private ArrayList splitTokens (String expr) {
		String[] exp = expr.split(""); // String -> String[] 
		ArrayList tokens = new ArrayList();
		int value = 0;
		String op = "";
		boolean flag = false;
		
		for(String c:exp) { // foreach문 (배열,ArrayList)
			if(c.equals(" ")) {
				continue;
			}
			
			if("0123456789".contains(c)) {
				value = value * 10 +  Integer.parseInt(c);
				flag = true;
			}else {
				if(flag) {
					tokens.add(value);
					value = 0;
				}
				flag = false;
				tokens.add(c);
			}
		}
		
		if(flag) {
			tokens.add(value);
		}
		return tokens;
	}
	
	/**
	 * 중위 표기식을 후위 표기식으로 변경합니다
	 * @param tokens 중위 표기식 ArrayList
	 * @return 후위 표기식 ArrayList
	 */
	private ArrayList infixToPostfix(ArrayList tokens) {
		ArrayList result = new ArrayList();
		HashMap<String, Integer> prec = new HashMap<String,Integer>();
		Stack opStack = new Stack();
		
		prec.put("*", 3);
		prec.put("/", 3);
		prec.put("+", 2);
		prec.put("-", 2);
		prec.put("(", 1);
		
		for(Object c:tokens) {
			if(c.equals("(")) {
				opStack.push(c);
			}else if(c.equals(")")) {
				while(!opStack.peek().equals("(")) {
					Object val = opStack.pop();
					if(!val.equals("(")) {
						result.add(val);
					}
				}
				opStack.pop();
			}else if(prec.containsKey(c)) {
				if(opStack.isEmpty()) {
					opStack.push(c);
				}else {
					if(prec.get(opStack.peek()) >= prec.get(c)) {
						result.add(opStack.pop());
						opStack.push(c);
					}else {
						opStack.push(c);
					}
				}
			}else {
				result.add(c);
			}
		}
		
		while(!opStack.isEmpty()) {
			result.add(opStack.pop());
		}
		
		return result;
	}
	
	/**
	 * 후위 표기식을 계산합니다.
	 * @param expr 후위 표기식 ArrayList
	 * @return 최종 계산 결과
	 */
	private int postFixEval(ArrayList expr) {
		Stack<Integer> valueStack = new Stack<Integer>();
		
		for(Object c:expr) {
			if(c instanceof Integer) {
				valueStack.push((Integer) c);
			}else if(c.equals("+")){
				int num1 = valueStack.pop();
				int num2 = valueStack.pop();
				
				valueStack.push(num2+num1);
			}else if(c.equals("-")){
				int num1 = valueStack.pop();
				int num2 = valueStack.pop();
				
				valueStack.push(num2-num1);
			}else if(c.equals("*")){
				int num1 = valueStack.pop();
				int num2 = valueStack.pop();
				
				valueStack.push(num2*num1);
			}else if(c.equals("/")){
				int num1 = valueStack.pop();
				int num2 = valueStack.pop();
				
				valueStack.push(num2/num1);
			}
		}
		
		int result = valueStack.pop();
		return result;
	}
	
	public int process(String expr) {
		ArrayList postfix = infixToPostfix(splitTokens(expr)); // 중위 표기식을 후위 표기식으로 변경
		int result = postFixEval(postfix); // 후위 표기 수식 계산
		return result;
	}
	
	public static void main(String args[]) {
		Scanner sc = new Scanner(System.in);
		String expr = sc.nextLine();
		Calc calc = new Calc();
		int result = calc.process(expr);
		System.out.println(result);
	}
}

File: Stack/연습문제/calc.py
from stack import ArrayStack

def split_tokens(expr):
    """
    문자열로 들어온 수식을 list로 분할
    :param expr:
    :return: 수식을 피연산자,연산자로 분할한 list
    """
    tokens = []
    val = 0
    val_processing = False

    for c in expr:
        if c == ' ':
            continue
        if c in '0123456789':
            val = val * 10 + int(c)
            val_processing = True
        else:
            if val_processing:
                tokens.append(val)
                val = 0
            val_processing = False
            tokens.append(c)

    if val_processing:
        tokens.append(val)

    return tokens

def infix_to_potfix(token_list):
    """
    들어온 중위 표현식을 후위 표현식으로 변경
    :param token_list:
    :return: 후위 표현식
    """
    prec = {
        '*': 3,
        '/': 3,
        '+': 2,
        '-': 2,
        '(': 1
    }

    opStack = ArrayStack()
    result = []

    for token in token_list:
        if isinstance(token, int):
            result.append(token)
        elif token == '(':
            opStack.push(token)
        elif token == ')':
            while opStack.peek() != '(':
                result.append(opStack.pop())
            opStack.pop()
        else:
            if not opStack.isEmpty():
                if prec[opStack.peek()] >= prec[token]:
                    result.append(opStack.pop())
                    opStack.push(token)
                else:
                    opStack.push(token)
            else:
                opStack.push(token)

    while not opStack.isEmpty():
        result.append(opStack.pop())

    print(result)
    return result

def post_fix_eval(expr):
    """
    후위 표현식으로 된 식을 계산
    :param expr:
    :return:
    """
    val_stack = ArrayStack()

    for token in expr:
        if isinstance(token, int):
            val_stack.push(token)
        elif token == '+':
            num1 = val_stack.pop()
            num2 = val_stack.pop()
            val_stack.push(num2+num1)
        elif token == '-':
            num1 = val_stack.pop()
            num2 = val_stack.pop()
            val_stack.push(num2-num1)
        elif token == '/':
            num1 = val_stack.pop()
            num2 = val_stack.pop()
            val_stack.push(num2/num1)
        elif token == '*':
            num1 = val_stack.pop()
            num2 = val_stack.pop()
            val_stack.push(num2*num1)

    return val_stack.pop()


print(post_fix_eval(infix_to_potfix(split_tokens("1*((1+2)*(3+4))"))))